test/parallel/test-assert-async.js

```diff
@@ -0,0 +1,48 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const promisify = require('util').promisify;
+const wait = promisify(setTimeout);
+
+// Ensure async support for assert.rejects() and assert.doesNotReject()
+/* eslint-disable prefer-common-expectserror, no-restricted-properties */
+
+async function expectsRejection(fn, expected) {
+  await assert.rejects(fn, expected);
+}
+async function expectsNoRejection(fn) {
+  await assert.doesNotReject(fn);
+}
+
+// catches synchronous error
+expectsRejection(() => {
+  assert.fail();
+}, common.expectsError({
+  code: 'ERR_ASSERTION',
+  type: assert.AssertionError,
+  message: 'Failed',
+  operator: undefined,
+  actual: undefined,
+  expected: undefined
+}));
+
+// awaits for async block and catches error
+expectsRejection(async () => {
+  await wait(10);
+  assert.fail();
+}, common.expectsError({
+  code: 'ERR_ASSERTION',
+  type: assert.AssertionError,
+  message: 'Failed',
+  operator: undefined,
+  actual: undefined,
+  expected: undefined
+}));
+
+// allows synchronous block
+expectsNoRejection(() => {});
+
+// awaits for async block
+expectsNoRejection(async () => {
+  await wait(10);
+});
```

Comment:
If I am not mistaken, we do not know if these really work because the function itself might theoretically execute sync.

So a better test out of my perspective would be:

```js
let promise;
let threw = false;
try {
  promise = assert.doesNotReject(async () => {
    await wait(10);
    assert.fail();
  });
} catch (err) {
  threw = true;
}
assert.strictEqual(threw, false);
assert.rejects(() => promise,
  common.expectsError({
    code: 'ERR_ASSERTION',
    type: assert.AssertionError,
    message: 'Got unwanted exception ... '
    ...
  })
);

Reply:
👌